lib/archethic/bootstrap/network_constraints.ex

```diff
@@ -0,0 +1,134 @@
+defmodule Archethic.Bootstrap.NetworkConstraints do
+  @moduledoc """
+  This GenServer make sure that Node load network constraints before any transaction validation.
+  Updates gen_addr constraints post bootstrap, if fails wait for node to get authorized
+  """
+
+  use GenServer, restart: :transient
+
+  alias Archethic.{
+    Reward,
+    SharedSecrets,
+    PubSub,
+    OracleChain,
+    TransactionChain
+  }
+
+  require Logger
+
+  def start_link(opts \\ []) do
+    GenServer.start_link(__MODULE__, opts, name: __MODULE__)
+  end
+
+  @spec init(any) ::
+          {:ok, %{node_shared_secrets: false, oracle: false, origin: false, reward: false}}
+  def init(_args) do
+    PubSub.register_to_node_update()
+    {:ok, %{reward: false, node_shared_secrets: false, origin: false, oracle: false}}
+  end
+
+  @spec persist_gen_addr :: :ok
+  def persist_gen_addr() do
+    GenServer.cast(__MODULE__, {:persist_gen_addr})
+  end
+
+  def handle_cast({:persist_gen_addr}, state) do
+    {:noreply, process_gen_addr(state)}
+  end
+
+  def handle_info(
+        {:node_update, _},
+        state
+      ) do
+    new_state = process_gen_addr(state)
+
+    Logger.info("Gen Addr: reward-#{new_state.reward}, origin: #{new_state.origin}",
+      nss: new_state.node_shared_secrets
+    )
+
+    if Enum.all?(new_state, fn {_k, v} -> v end) do
+      Logger.info("Stopping Gen Addr loader gracefully")
+      {:stop, :normal, new_state}
+    else
+      {:noreply, state}
+    end
+  end
+
+  def process_gen_addr(%{
+        reward: reward_gen_addr?,
+        node_shared_secrets: nss_gen_addr?,
+        origin: origin_gen_addr?,
+        oracle: oracle_gen_addr?
+      }) do
+    %{
+      reward: persist(:reward, reward_gen_addr?),
+      node_shared_secrets: persist(:node_shared_secrets, nss_gen_addr?),
+      origin: persist(:origin, origin_gen_addr?),
+      oracle: persist(:oracle, oracle_gen_addr?)
+    }
+  end
+
+  def persist(:reward, true), do: true
+  def persist(:origin, true), do: true
+  def persist(:oracle, true), do: true
+  def persist(:node_shared_secrets, true), do: true
+
+  def persist(:oracle, false) do
+    try do
+      {next_summary_date_gen_addr, next_summary_date} = OracleChain.derive_current_gen_addr()
+      OracleChain.put_gen_addr(next_summary_date_gen_addr, next_summary_date)
+      true
+    rescue
+      _error -> false
+    end
+  end
+
+  def persist(:reward, false) do
+    try do
+      case TransactionChain.list_addresses_by_type(:mint_rewards)
+           |> Enum.at(0) do
```

Comment:
Here you can do some improvement. As we only want the first address we don't need to wait for all the addresses to be fetched.
`TransactionChain.list_addresses_by_type` will return a Stream and `Enum.at(0)` will wait for all the Stream to be finished before taking the element 0.
You can add `Stream.take(1)` before the Enum. This will stop the Stream after the first element !
```
TransactionChain.list_addresses_by_type(:mint_rewards)
|> Stream.take(1)
|> Enum.at(0)
```

Reply:
nice